### PR TITLE
Add auto generation of __typename field

### PR DESCRIPTION
**Pull request by sav007**
Part of #451

### CODE REVIEW

File: apollo-compiler/src/main/kotlin/com/apollographql/android/compiler/ir/Field.kt [hunk 1]
```diff
@@ -12,21 +12,24 @@ data class Field(
...
+    val args: List<Map<String, Any>>? = null,
     val isConditional: Boolean = false,
-    val fields: List<Field>?,
-    val fragmentSpreads: List<String>?,
-    val inlineFragments: List<InlineFragment>?,
-    val description: String?
+    val fields: List<Field>? = null,
+    val fragmentSpreads: List<String>? = null,
+    val inlineFragments: List<InlineFragment>? = null,
+    val description: String? = null
```
Comment (marwanad) [inline apollo-compiler/src/main/kotlin/com/apollographql/android/compiler/ir/Field.kt]: what made those null initializations necessary?

File: apollo-compiler/src/main/kotlin/com/apollographql/android/compiler/ir/Fragment.kt [hunk 2]
```diff
@@ -10,16 +10,23 @@ import javax.lang.model.element.Modifier
...
     val inlineFragments: List<InlineFragment>,
     val fragmentsReferenced: List<String>
 ) : CodeGenerator {
+
   /** Returns the Java interface that represents this Fragment object. */
   override fun toTypeSpec(context: CodeGenerationContext): TypeSpec =
-      SchemaTypeSpecBuilder(formatClassName(), fields, fragmentSpreads, inlineFragments, context)
+      SchemaTypeSpecBuilder(
+          typeName = formatClassName(),
+          fields = listOf(Field("__typename", "__typename", "String!")) + fields,
```
Comment (BenSchwab) [inline apollo-compiler/src/main/kotlin/com/apollographql/android/compiler/ir/Fragment.kt]: nvm, I guess that is just to the query. But this to actually add it as a field. 👍

Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/android/compiler/ir/Fragment.kt]: yes, it auto adds to query but we still need to generate field for it

File: apollo-compiler/src/main/kotlin/com/apollographql/android/compiler/ir/Field.kt (hunk 1 above)
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/android/compiler/ir/Field.kt]: `Field("__typename", "__typename", "String!")` this stuff

**Pull request merged**